### ISSUE TITLE
Failure to correctly set ids.store-name=titan_ids can corrupt a Titan Cassandra

### ISSUE THREAD

**Issue by smcquillan** [JanusGraph/janusgraph#418]
I've been running into issues while connecting JanusGraph 0.1.2-SNAPSHOT to instances of Cassandra which have graphs previously populated by Titan 1.0.0. Initially things looked to be working fine as JanusGraph could read the data but things began to go wrong when I created a new property with something like the following:

```
JanusGraphManagement mgmt = g.openManagement();
mgmt.makePropertyKey("newProperty").dataType(type).make();
```

The `newProperty` which had a data type of `String` got created fine. However, I later discovered that one of my existing properties which had previously been of type `Long` had now changed to type `String` causing many of our application tests to fail and effectively **the graph within Cassandra was now corrupted**.

After digging around a bit found issue #228 about ID_STORE_NAME table renamed from Titan and thought that must be where I was going wrong. Looking in Cassandra I saw that in addition to the existing `titan_ids` a new `janusgraph_ids` had been created and that confirmed it. But that wasn't the whole story because just adding the following to the configuration when connecting:

```
graph.titan-version=1.0.0
ids.store-name=titan_ids
```

Works fine when connecting to an empty Cassandra, but if you connect to one which had previously been populated by Titan you'll get the following:

```
WARN [2017-07-18 12:03:43,880] [main] o.j.g.c.GraphDatabaseConfiguration -  Local setting ids.store-name=titan_ids (Type: GLOBAL_OFFLINE) is overridden by globally managed value (janusgraph_ids).  Use the ManagementSystem interface instead of the local configuration to control this setting. 

```

The message is questionable because the implicit value of `ids.store-name` in the Cassandra being connected to is actually `titan_ids` so we're not overridding anything. And worse still if you don't notice the message and then go on to change the schema you'll risk corrupting the graph because the changes will to into `janusgraph_ids` instead. I worked around the issue by using the  ManagementSystem interface as the warning suggested:

```
mgmt = g.openManagement();
mgmt.set(ConfigElement.getPath(GraphDatabaseConfiguration.IDS_STORE_NAME), "titan_ids");
mgmt.commit();
```

Even this had its complications as you need to bring down the cluster and connect with just a single JanusGraph / Titan as any other connected instances will prevent the change from being applied.

Raising this as an issue because while the [Upgrading from Titan](https://github.com/JanusGraph/janusgraph/blob/1239aad95a8bb96872303a95f7dedc17056df799/UPGRADE.asc) does mention `ids.store-name` it implies that setting it in the JanusGraph properties file is sufficient which in my experience it was not. Ideally I'd like to see it settable from the properties file, as that link suggests, though perhaps only allowed if `graph.titan-version=x.x.x` has also been set.

**Comment by pluradj** [comment]
Thanks for reporting @smcquillan -- I was able to see the `WARN` behavior, but I didn't manage to reproduce the graph corruption. I can understand how it could happen though because some of the ids are stored in `titan_ids` while newer ones are stored in `janusgraph_ids`.

Based on your write up, it sounds like you are suggesting 2 options.

First option is to update the migration docs with the extra steps to handle the `GLOBAL_OFFLINE` property update:
1. shutting down all connections to the graph
2. updating the `ids.store-name` property with `titan_ids` in the stored graph configuration via the Management System
3. restarting connections to the graph
After all of this is done, `janusgraph_ids` table will still exist in the keyspace, but it will not be used. Perhaps this table should be deleted also?

Second option is for additional logic prior to opening the storage db connection to look for `graph.titan-version` and `ids.store-name` config properties, which would indicate a migrated Titan instance. This should be a more seamless experience, and I think it would avoid creating the unused `janusgraph_ids` table.

Are you interested in taking this one on?

**Comment by smcquillan** [comment]
@pluradj good summary and yes for the first option I did still notice that, as you suspected, the `janusgraph_ids` table had still been created as I saw a directory of that name in the Cassandra data directory. However, unlike before when I didn't change  `ids.store-name` it was empty after I created the new proprety and so I assume had not been used.

I favour the second option and while debugging the issue I'd got as far as identifying the [GraphDatabaseConfiguration](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java#L1344) constructor as the place where the value of `ids.store-name` in `localConfig` had got *lost* . So perhaps its in here that the extra logic for `graph.titan-version` and `ids.store-name` config properties would fit.

If people in agreement that the second option is the best way forward, I don't mind taking this one on.

**Comment by pluradj** [comment]
@amcp put the fix in for #228 so it'd be good to hear his feedback

**Comment by amcp** [comment]
You can check for the presence of ids.store-name in config and throw an exception if not present in certain circumstances. Is it possible to check for the presence of ids.store-name config at startup, after we have read the database version from system_properties?

**Comment by smcquillan** [comment]
@amcp did some testing and it does look like its possible to check for the presence of `ids.store-name` at startup after reading form the database. I added the second two lines of following to the GraphDatabaseConfiguration constructor to print it out:
```
        KCVSConfiguration kcvsConfig=Backend.getStandaloneGlobalConfiguration(storeManager,localbc);
        
        ReadConfiguration kcvsReadConfig = kcvsConfig.asReadConfiguration();
        System.out.println("KCVS: IDS_STORE_NAME: " + kcvsReadConfig.get("ids.store-name", String.class));
```
And the results when reading from a Cassandra previously populated by Titan ( and an empty Cassandra ) was the following: 
```
KCVS: IDS_STORE_NAME: null
```
Whereas the results reading from a Cassandra previously populated by JanusGraph are:
```
KCVS: IDS_STORE_NAME: janusgraph_ids
```
From this it looks like we're in a good position to say that if the valid from the KCVS store is 'null' and `graph.titan-version=x.x.x` and `ids.store-name=titan_ids` then allow it to be set from the configuration file.  If anyone should then subsequently try and change it to a different value that'd be caught as after that first time JanusGraph will have written the new value to Cassandra and it would no longer be null in the KCVS store.

**Comment by amcp** [comment]
@smcquillan I sort-of agree with the approach. On second thought, I'm not sure I like the fact that we have to bring down a Cassandra cluster to change this value. Seems like we could just set it locally, and just use the titan_ids table as usual. I can foresee a need to do a wave restart of hosts in the cassandra cluster and change the nodes one by one. Maybe the answer is, don't do any of these tasks, and just make ids.store-name a LOCAL configuration.

Furthermore, to avoid data inconsistency like what you experienced in your tests, we could do a startup check in the place where we can read the global config from management. Throw an exception and abort if:
1. graph.titan-version is set and is 1.0.0 AND
2. false == storeManager.storeExists(janusgraph_ids) AND
3. storeManager.storeExists(titan_ids) AND
4. ids.store-name is not set in the local configuration,

**Comment by smcquillan** [comment]
@amcp I'm in complete agreement about trying to avoid bringing down a Cassandra cluster to change this value and this is what I want to avoid. I think the problem boils down the fact that JanusGraph has ids.store-name in the GLOBAL configuration and persists this to Cassandra whereas Titan does not.

Bit reluctant to move ids.store-name to LOCAL given that its currently GLOBAL, would that not affect existing JanusGraph users? That said, can we leave it GLOBAL if we, as you suggest, raise an exception for the following:

1. graph.titan-version is set and is 1.0.0 AND
2. false == storeManager.storeExists(janusgraph_ids) AND
3. storeManager.storeExists(titan_ids) AND
4. ids.store-name is not set in the local configuration,

But case where for 4 the `ids.store-name` was set in the local configuration and matched the value from 3. We update the global configuration to include the value with had previously been missing (because Titan didn't set it).

**Comment by smcquillan** [comment]
I've had some success with the following logic:

1. `graph.titan-version` is set and is '1.0.0'
2. `ids.store-name` is set in LOCAL and equal to 'titan_ids'
3. `ids.store-name` is **not** set in GLOBAL ( as read from KCVSConfiguration )

If all the above are true set `ids.store-name` in the  [overwrite](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java#L1350) modifiable configuration to be `titan_ids`. This then prevents the global configuration inheriting the default value of `janusgraph_ids` but only in the case where `ids.store-name` had not already been set.

**Comment by smcquillan** [comment]
@amcp @pluradj  I've pushed some changes up to a local branch [titanIdStoreIssue418](https://github.com/JanusGraph/janusgraph/compare/master...smcquillan:titanIdStoreIssue418) which fixes the issue in my environment. 

The majority of the changes being in the `checkBackwardCompatibilityWithTitan` method using the logic in the comment above. There's one other change to check if default global value for the ID store is going to be overwritten (without this check it works but you get the warning in the log about overriding a globally managed value).

It will now throw an exception and abort if `graph.titan-version` has been set and `ids.store-name` was not also set to the value `titan_ids` which should help preserver data integrity.

Does this approach look OK and shall I go ahead and raise a pull request?

**Comment by pluradj** [comment]
I gave it a quick test, and it worked how I expected. Opening the pull request sounds great to me.

**Comment by jerryjch** [comment]
Fixed via PR #445

**Comment by rafatuita** [comment]
Hi guys,
I just upgraded my janus to 0.2.0 and found out that this change broke my system. 
in this line of code:
`
boolean localTitanConfigured = localbc.get(TITAN_COMPATIBLE_VERSIONS) != null;
`
Apparently, my TITAN_COMPATIBLE_VERSIONS are in the Global and not on Local config, therefore I'm getting the following.

I'm sending a PR with a try/catch to fix this ok? @smcquillan @pluradj @amcp @jerryjch @sjudeng 

`java.lang.IllegalStateException: Need to set configuration value: root.graph.titan-version
	at com.google.common.base.Preconditions.checkState(Preconditions.java:197)
	at org.janusgraph.diskstorage.configuration.ConfigOption.get(ConfigOption.java:230)
	at org.janusgraph.diskstorage.configuration.BasicConfiguration.get(BasicConfiguration.java:69)
	at org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration.checkBackwardCompatibilityWithTitan(GraphDatabaseConfiguration.java:1536)
	at org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration.<init>(GraphDatabaseConfiguration.java:1430)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:164)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:133)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:123)
	at org.janusgraph.core.JanusGraphFactory$Builder.open(JanusGraphFactory.java:264)
	at com.wisgle.microservices.dao.oltp.factory.GraphFactory.getJanusGraph(GraphFactory.java:40)
	at com.wisgle.microservices.dao.oltp.dao.impl.CommonsDaoImpl.init(CommonsDaoImpl.java:50)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:365)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:310)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:133)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:408)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1570)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:545)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:207)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1214)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1054)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1019)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:566)
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:349)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1214)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:543)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:207)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1214)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1054)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1019)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:566)
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:349)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1214)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:543)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:776)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:861)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:541)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:122)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:369)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:313)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1185)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1174)
	at com.wisgle.microservices.dao.oltp.Application.main(Application.java:56)
java.lang.IllegalStateException: Need to set configuration value: root.graph.titan-version
	at com.google.common.base.Preconditions.checkState(Preconditions.java:197)
	at org.janusgraph.diskstorage.configuration.ConfigOption.get(ConfigOption.java:230)
	at org.janusgraph.diskstorage.configuration.BasicConfiguration.get(BasicConfiguration.java:69)
	at org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration.checkBackwardCompatibilityWithTitan(GraphDatabaseConfiguration.java:1536)
	at org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration.<init>(GraphDatabaseConfiguration.java:1430)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:164)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:133)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:123)
	at org.janusgraph.core.JanusGraphFactory$Builder.open(JanusGraphFactory.java:264)
	at com.wisgle.microservices.dao.oltp.factory.GraphFactory.getJanusGraph(GraphFactory.java:40)
	at com.wisgle.microservices.dao.oltp.dao.impl.CommonsDaoImpl.init(CommonsDaoImpl.java:50)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:365)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:310)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:133)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:408)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1570)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:545)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:207)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1214)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1054)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1019)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:566)
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:349)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1214)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:543)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:207)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1214)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1054)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1019)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:566)
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:349)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1214)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:543)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:776)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:861)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:541)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:122)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:369)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:313)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1185)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1174)
	at com.wisgle.microservices.dao.oltp.Application.main(Application.java:56)
java.lang.IllegalStateException: Need to set configuration value: root.graph.janusgraph-version
	at com.google.common.base.Preconditions.checkState(Preconditions.java:197)
	at org.janusgraph.diskstorage.configuration.ConfigOption.get(ConfigOption.java:230)
	at org.janusgraph.diskstorage.configuration.BasicConfiguration.get(BasicConfiguration.java:69)
	at org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration.checkBackwardCompatibilityWithTitan(GraphDatabaseConfiguration.java:1536)
	at org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration.<init>(GraphDatabaseConfiguration.java:1430)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:164)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:133)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:123)
	at org.janusgraph.core.JanusGraphFactory$Builder.open(JanusGraphFactory.java:264)
	at com.wisgle.microservices.dao.oltp.factory.GraphFactory.getJanusGraph(GraphFactory.java:40)
	at com.wisgle.microservices.dao.oltp.dao.impl.CommonsDaoImpl.init(CommonsDaoImpl.java:50)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:365)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:310)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:133)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:408)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1570)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:545)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:207)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1214)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1054)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1019)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:566)
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:349)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1214)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:543)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:207)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1214)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1054)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1019)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:566)
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:349)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1214)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:543)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:776)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:861)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:541)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:122)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:369)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:313)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1185)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1174)
	at com.wisgle.microservices.dao.oltp.Application.main(Application.java:56)
`
